File: src/main/java/org/cloudfoundry/credhub/util/TimeModuleFactory.java
package org.cloudfoundry.credhub.util;

import java.io.IOException;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import static java.time.format.DateTimeFormatter.ofPattern;

final public class TimeModuleFactory {

  private static final DateTimeFormatter TIMESTAMP_FORMAT = ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'");

  private TimeModuleFactory() {
    super();
  }

  public static JavaTimeModule createTimeModule() {
    final JavaTimeModule javaTimeModule = new JavaTimeModule();

    javaTimeModule.addSerializer(Instant.class, new JsonSerializer<Instant>() {
      @Override
      public void serialize(final Instant value, final JsonGenerator gen, final SerializerProvider serializers)
        throws IOException {
        gen.writeString(ZonedDateTime.ofInstant(value, ZoneId.of("UTC")).format(TIMESTAMP_FORMAT));
      }
    });

    return javaTimeModule;
  }
}

File: components/time-support/src/main/kotlin/org/cloudfoundry/credhub/util/TimeModuleFactory.kt
package org.cloudfoundry.credhub.util

import tools.jackson.core.JsonGenerator
import tools.jackson.databind.JacksonModule
import tools.jackson.databind.SerializationContext
import tools.jackson.databind.ValueSerializer
import tools.jackson.databind.module.SimpleModule
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter.ofPattern

class TimeModuleFactory private constructor() {
    companion object {
        private val TIMESTAMP_FORMAT = ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")

        fun createTimeModule(): JacksonModule {
            val module = SimpleModule("CredhubTimeModule")

            module.addSerializer(
                Instant::class.java,
                object : ValueSerializer<Instant>() {
                    override fun serialize(
                        value: Instant,
                        gen: JsonGenerator,
                        serializers: SerializationContext,
                    ) {
                        gen.writeString(ZonedDateTime.ofInstant(value, ZoneId.of("UTC")).format(TIMESTAMP_FORMAT))
                    }
                },
            )

            return module
        }
    }
}
